Segment try cleanup

liujing2 opened the pull request:
Hi @mcastelino,

I tried something today before I can know your detailed plan. Although this is WIP, the first patch might be useful. 
I am not sure if your plan before really cleanup is all about memory allocation? If so, I think I can continue. Otherwise, I need wait for your email :)

Oops, it seems no update in your branch, so could you only look at the last two? Thanks!

Commit:
hw: i386: virt: Do not include pc.h

We don't need the PC machine type header from the virt code.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: i386: Make the hotpluggable memory size property more generic

This property is currently defined under i386/pc while it only describes
a region size that's eventually fetched from the AML ACPI code.

We can make it more generic and shareable across machine types by moving
it to memory-device.h instead.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: i386: Move PCI host definitions to pci_host.h

The PCI hole properties are not pc or i386 specific. They belong to the
PCI host header instead.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
acpi: Remove aml-build i386/pc dependency

Now that the PCI host properties are moved away from pc.h, we can remove
this header dependency from aml-build.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: acpi: Remove PC dependency from the CPU hotplug code

There are no reasons why we'd need to include the PC machine type header
for doing generic ACPI CPU hotplug.

Here we replace pc.h with pci/pci.h and change the legacy AML code to
only depend on the generic AcpiConfiguration structure instead of the PC
machine type state.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: acpi: Remove pc header inclusion from the PCI hotplug code

It is simply not needed.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: i386: Move all PCI definitions out of pc.h

They're generic, at worst x86 specific, definitions and should belong to
a more generic header: i386/pci.h.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: i386: Remove the pc header dependency from the x86 code

The x86 generic API should be dependent on the PC machine type.
To break that dependency, we have to make a couple of x86 APIs
independent from the pc machine type API.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: i386: Remove the pc header dependency from the cpu code

It's only a matter of moving the compat APIC boolean to the correct
header file (apic.h).

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: i386: Remove the pc header dependency from the fw code

It's only a matter of moving the FW IO base definition to the correct
header file (fw.h).

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: i386: Remove the pc header dependency from the kvm IOAPIC code

We need to move the GSI state structure definition to the ioapic header.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: i386: Make the SMM machine property generic

SMM is an x86 property, not a PC machine type one.
This will help making the SMM enablement check PC machine type agnostic.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: i386: Move the no_hpet definition out of pc.h

The no_hpet toggle is x86 specific, not PC specific.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
target: i386: Remove the pc header dependency from the kvm code

There is no need to make the SMM enablement check pc machine type
dependent. We can simply request for the SMM property instead.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: i386: Export the PC machine class save_tsc_khz field

Export it as a read-only property. We only need it to be able to fetch
it from other places, without having to depend on the PC machine type
definitions.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
target: i386: Remove the pc header dependency from the machine code

Instead of relying on PC machine type specific macros, we fetch the TSC
rate migration setting through a property.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
target: i386: Remove the pc header dependency from the monitor code

It is simply not needed.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: i386: Remove unneeded chipset header from the ACPI build code

It's simply not needed anymore.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: i386: Factorize madt cpu entry routine

Both the pc and virt machine types share the same madt cpu entry
routine, so we can share it.
By doing so, we are also able to remove the pc_madt_cpu_entry.c pc
header dependency.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
tools: Disable BZIP2 for all builds

It's only needed for the DMG block compression code and we disable DMG
support completely.

Fixes #133

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
pci-host: Change the way to get pci_host in aml_build.c

Currently, Qemu assumes one PCI Host Bridge on a platform, so it statically
gets pci_host by resolving the string path. This patch changes the way by
adding AcpiPciBus to cast the PCIBus to pci_host, which will help multi-segment
work easier and cleaner in next step.

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

mcastelino commented on include/hw/i386/virt.h:
```diff
@@ -46,7 +47,7 @@ typedef struct {
     qemu_irq *gsi;
 
-    PCIBus *pci_bus;
+    PCIBus *pci_bus[MAX_PCI_SEGMENT];
```
We should make this a list vs an array. That will give us more flexibility and the ability to add to this list.

mcastelino commented on hw/i386/virt/virt.c:
```diff
@@ -185,6 +189,11 @@ static void virt_machine_state_init(MachineState *machine)
 
+    //TODO: This needs some property to specify
+    //Can be put into VirtMachineState
+    int segment_total_nr = 2;
```
I am still trying to see if we should have the ability to add as many segments as we need on the command line vs setting it to a predetermined number of segments. Allowing the user to specify the number of segments will make this looks more like a pcie_pxb along the lines of 

```
-device pxb-pcie,id=pcie.1,bus_nr=x[,numa_node=y][,addr=z]
```

mcastelino commented on hw/acpi/aml-build.c:
```diff
@@ -1673,12 +1675,13 @@ void acpi_get_pci_holes(Range *hole, Range *hole64)
 
 
-bool acpi_get_mcfg(AcpiMcfgInfo *mcfg)
+bool acpi_get_mcfg(AcpiMcfgInfo *mcfg, AcpiPciBus *acpi_pci_bus)
```
We need to do this differently. This should be sent in a list and we should build the table from the list.

mcastelino reviewed (commented):
@liujing2 added some feedback.

Commit:
Add "segment_nr" property for VirtMachineState

Users can specify the number of pci host bridge when launching
guest by "-machine virt,segment-nr=N" property.

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

liujing2 commented:
@mcastelino  I updated with some more work today. 
Still conflict but I think we can read the patches.

Commit:
BUG fix for upstream: Fix checking when setting property

Qemu uses a static bool to avoid the property being set twice but it doesn't
check the owner. This might be because originally there is only one PCI Host
Bridge. For different PCI Host Bridge, ACPI_PCIHP_PROP_BSEL should be
added to each root bus.

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
MCFG: Make MCFG build more generic

ACPI builds MCFG for single PCI host before because QEMU assumes only
one PCI host on pc and q35 platform. Making MCFG build more generic
by giving the PCI host list. Then MCFG table can include multiple
allocations to reside multiple PCI host information.

NOTE: One case needs discuss:
If one of the PCI hosts doesn't have MCFG_BASE address, just return a
NULL MCFG table, does this make sense?

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
Add segment_nr into AcpiConfiguration

To enable ACPI build tables with multi-segment, adding the variable
to AcpiConfiguration and will use it later.

On PC and q35 platform, we set it to one.

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
pci-lite: Make PCI host initialization for multiple host bridge

Once we want to create multiple PCI host for virt platform, we need
use pci_lite_init to help create multiple PCI host and root bus.

Later, we will set acpi related functions for PCI root bus list.

Use 0.pci.0 as the name for the first PCI host.

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
[WIP]pcihp: Make ACPI PCIHP for multiple root bus

Set different names for ACPI PCIHP memory region
This doesn't work now.

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
Add property check before set

Qemu assumes only one PCI host so it doesn't check the property
adding of ACPI_PCIHP_IO_BASE_PROP. For virt platform, when multiple
PCI segment is used and ACPI PCI hotplug is enabled, it needs check
if this property is already set before.

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
Build ACPI tables for multiple root bus

Make PCI host as a list when build hardware reduced ACPI for virt platform,
so that it builds DSDT and MCFG referring to the amount of PCI segment.

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
[WIP]Removing hardcoding address for each PCI host

The virt platform decides which address should be allocated for each
PCI segment by a simple logic.

Reduce the low memory size to 1G so that we reserve 3G for PCI.

0x40000000~0x9fffffff: MCFG for each segment
0xa0000000~IO_APIC_DEFAULT_ADDRESS: PCI HOLE

This now doesn't work. Still need improvement.

Question:
1. We can't enable MCFG above 4G, so only use below 4G
2. Is IOAPIC and LAPIC address space useful? Could we remove that?
3. If each MCFG use 256M, we could only set 4 PCI segment.

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

mcastelino commented on hw/i386/virt/memory.c:
```diff
@@ -25,7 +25,7 @@
 #include "cpu.h"
 
-#define VIRT_LOWMEM 0x80000000
+#define VIRT_LOWMEM 0x40000000
```
@liujing2 if you change this, you need to make a corresponding change in OVMF where the value is hardcoded in a GUID.

mcastelino replied:
gEfiMdePkgTokenSpaceGuid.PcdPciExpressBaseAddress

mcastelino replied:
@liujing2 additionally OVMF ignores the E820 reservations. Hence we have to send in the number of segments also via fw_cfg. This will allow OVMF to setup the EFI memory map correctly so that each segment's range is added properly.

mcastelino commented on include/hw/i386/virt.h:
```diff
@@ -76,4 +76,11 @@ MemoryRegion *virt_memory_init(VirtMachineState *vms);
 
 DeviceState *virt_acpi_init(qemu_irq *gsi, PCIBus **pci_bus, uint16_t segment_total);
 
+#define PCI_HOST_PCIEXBAR_BASE 0x40000000
```
@liujing2 ideally I want to pass this via fw_cfg so that the firware does not hardcode this value.

liujing2 replied:
OK, let me do that.

mcastelino commented on hw/pci-host/pci_lite.c:
```diff
@@ -230,15 +225,23 @@ PCIBus *pci_lite_init(MemoryRegion *address_space_mem,
+    mcfg_base = PCI_HOST_PCIEXBAR_BASE + nr * PCI_HOST_PCIEXBAR_SIZE;
+    pci_hole_base = PCI_HOST_PCI_HOLE_BASE + nr * PCI_HOST_PCI_HOLE_SIZE;
+    pci_hole_end = pci_hole_base + PCI_HOST_PCI_HOLE_SIZE;
+    /* TODO: Do we still need IOAPIC and LAPCI spaces since using kernel_irqchip? */
```
@liujing2 yes. we still need to keep these addresses as they are reserved. The 8MB range above is also mapped for MSI.

liujing2 replied:
OK, got it. KVM uses the address, I see.

mcastelino commented on hw/acpi/reduced.c:
```diff
@@ -185,9 +185,16 @@ static void acpi_reduced_build(MachineState *ms, AcpiBuildTables *tables, AcpiCo
+    mcfg = g_new0(AcpiMcfgInfo, 1);
+    mcfg->segment_total = conf->segment_nr;
+
+    /* TODO: What if one of the multi pci-host doesn't have mcfg base?
```
@liujing2 not sure I understand this comment. Can you explain what the issue is out here.

liujing2 replied:
@mcastelino sure. Originally if the pci-host doesn't set PCIE_HOST_MCFG_BASE, acpi_get_mcfg will return false so it doesn't build MCFG. In our case, we have more than one pci-host, if one of them doesn't set the BASE, do we build MCFG or not?

mcastelino commented:
@liujing2 I rebased your code so that we can review more easily. https://github.com/mcastelino/nemu/pull/6

Also in OVMF you need a small change for all of this to work

```
diff --git a/OvmfPkg/OvmfPkgX64.dsc b/OvmfPkg/OvmfPkgX64.dsc
index 0aeb386650..d13fc80695 100644
--- a/OvmfPkg/OvmfPkgX64.dsc
+++ b/OvmfPkg/OvmfPkgX64.dsc
@@ -505,7 +505,7 @@
   #
   # On Q35 machine types that QEMU intends to support in the long term, QEMU
   # never lets the RAM below 4 GB exceed 2 GB.
-  gEfiMdePkgTokenSpaceGuid.PcdPciExpressBaseAddress|0x80000000
+  gEfiMdePkgTokenSpaceGuid.PcdPciExpressBaseAddress|0x40000000^M

 !ifdef $(SOURCE_DEBUG_ENABLE)
   gEfiSourceLevelDebugPkgTokenSpaceGuid.PcdDebugLoadImageMethod|0x2
```

I will add a few more changes to OVMF to get multiple segments to work.
